sdks/go/pkg/beam/core/graph/coder/panes.go

```diff
@@ -0,0 +1,143 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package coder
+
+import (
+	"io"
+
+	"github.com/apache/beam/sdks/go/pkg/beam/core/typex"
+	"github.com/apache/beam/sdks/go/pkg/beam/core/util/ioutilx"
+)
+
+const (
+	FIRST     int = 0
+	ONE_INDEX int = 1
+	TWO_INDEX int = 2
+)
+
+func chooseEncoding(v typex.PaneInfo) int {
+	if (v.Index == 0 || v.NonSpeculativeIndex == 0) || v.Timing == typex.UNKNOWN {
+		return FIRST
+	} else if v.Index == v.NonSpeculativeIndex || v.Timing == typex.EARLY {
+		return ONE_INDEX
+	} else {
+		return TWO_INDEX
+	}
```

Comment:
Prefer using a switch statment for things like this instead of relying on if-else ladders.
```suggestion
	switch {
	case v.Index == 0 || v.NonSpeculativeIndex == 0) || v.Timing == typex.UNKNOWN:
 		return FIRST
	case v.Index == v.NonSpeculativeIndex || v.Timing == typex.EARLY:
		return ONE_INDEX
	default:
		return TWO_INDEX
	}
```

Reply:
As a result, we can probably just move these conditions into the encoder, rather than relating them through constants, and have a comment that explains their differences, which you're using as a the constant name.